cpp/src/structs/utilities.cpp

```diff
@@ -288,59 +370,107 @@ std::tuple<cudf::column_view, std::vector<rmm::device_buffer>> push_down_nulls(
-                       null_count);
+      ret_nullable_data.new_null_masks.push_back(std::move(new_mask));
+      return std::pair{
+        reinterpret_cast<bitmask_type const*>(ret_nullable_data.new_null_masks.back().data()),
+        null_count};
     }();
 
-    return cudf::column_view(
-      child.type(),
-      child.size(),
-      child.head(),
-      new_child_mask,
-      null_count,
-      child.offset(),
-      std::vector<cudf::column_view>{child.child_begin(), child.child_end()});
+    return column_view(child.type(),
+                       child.size(),
+                       child.head(),
+                       new_child_mask,
+                       null_count,
+                       child.offset(),
+                       std::vector<column_view>{child.child_begin(), child.child_end()});
   };
 
-  auto child_begin =
-    thrust::make_transform_iterator(thrust::make_counting_iterator(0), rewrite_child_mask);
-  auto child_end = child_begin + structs_column.num_children();
+  auto const child_begin =
+    thrust::make_transform_iterator(thrust::make_counting_iterator(0), child_with_new_mask);
+  auto const child_end = child_begin + structs_view.num_children();
+  auto ret_children    = std::vector<column_view>{};
 
-  auto ret_children = std::vector<cudf::column_view>{};
   std::for_each(child_begin, child_end, [&](auto const& child) {
-    auto [processed_child, backing_buffers] = push_down_nulls(child, stream, mr);
-    ret_children.push_back(processed_child);
-    ret_validity_buffers.insert(ret_validity_buffers.end(),
-                                std::make_move_iterator(backing_buffers.begin()),
-                                std::make_move_iterator(backing_buffers.end()));
+    auto [processed_child, child_nullable_data] = push_down_nulls_no_sanitize(child, stream, mr);
+    ret_children.emplace_back(std::move(processed_child));
+    ret_nullable_data.emplace_back(std::move(child_nullable_data));
   });
 
   // Make column view out of newly constructed column_views, and all the validity buffers.
 
-  return std::make_tuple(column_view(input.type(),
-                                     input.size(),
-                                     nullptr,
-                                     input.null_mask(),
-                                     input.null_count(),  // Alternatively, postpone.
-                                     input.offset(),
-                                     ret_children),
-                         std::move(ret_validity_buffers));
+  return std::pair{column_view(input.type(),
+                               input.size(),
+                               nullptr,
+                               input.null_mask(),
+                               input.null_count(),  // Alternatively, postpone.
+                               input.offset(),
+                               ret_children),
+                   std::move(ret_nullable_data)};
```

Comment:
The purpose of such functions is to fix a `column_view` or `table_view` for intermediate computation, such as sorting, so we only need to generate a new `column_view`/`table_view` with all nulls are empty. If we return new `column`(s) from the input `column_view`(s), we always have to copy (a lot of) data from the input to the output.

The current approach is trying to avoid copying as much as possible. Only in the cases when we want to purge non-empty nulls then we will do copying.